### ISSUE TITLE
Split tools into individual plugins

### ISSUE THREAD

**Issue by Manishearth** [rust-diplomat/diplomat#53]
It would be nice if each plugin were a separate tool, perhaps callable via `diplomat-tool` but with a general interface

**Comment by Manishearth** [comment]
Ideally we can have common flags for `diplomat-tool` and each tool can take in more flags, something like `diplomat-tool --json-dependency=foo.json  --json-dependency=bar.json c --flags-for-c --more-flags-for-c`, where `c` means that `diplomat-tool` calls `diplomat-tool-c`

**Comment by CBenoit** [comment]
FYI, I’m thinking about giving a shot to this.
I’ll start by extracting the dotnet backend to a separate crate called `diplomat-tool-dotnet`.

`diplomat-tool` will parse the Rust source code into an AST, lower it to diplomat HIR and serializes the result into JSON using `serde`.
The backend executable is then spawned as a subprocess and the serialized payload is wrote to into subprocess’s stdin.
The backend then reads stdin, parses the input back into the original HIR and do its job.
Other formats besides JSON could be used / supported as well, but since JSON is one of the most widely available format, in theory we could even support backends not written in Rust, I figured it would be a good candidate.

If the serialized result could be printed to stdout, this:
```
diplomat-tool [--tool-flags ..] dotnet [--dotnet-flags ..]
```
could be roughly equivalent to this:
```shell
diplomat-tool [--tool-flags ..] | diplomat-tool-dotnet [--dotnet-flags ..]
```

By the way, I was thinking that instead of names like `diplomat-tool-dotnet`, "plugins" crates could drop "tool" and be named `diplomat-c`, `diplomat-cpp`, `diplomat-dotnet`… since those are really more like "diplomat backends". It would make sense to me at least. I know this is bikeshedding, but I’m throwing the idea while I’m at it.

**Comment by Manishearth** [comment]
Hmm. Though I'm probably going to add spans/etc to the AST eventually that may cause problems for serializability.

I think a good first step would be to do that kind of split as a library thing without the JSON intermediary.

**Comment by CBenoit** [comment]
> Though I'm probably going to add spans/etc to the AST eventually that may cause problems for serializability.

This is a good thing, are you thinking about embedded `&str` slices into the AST to refer to the original input?

Just to clarify, I was thinking about making the HIR serializable only, not the AST. That being said, embedded some sort of spans to allow the backend to show helpful diagnostics is still desirable. Maybe this can be solved if the span holds the following:
- start index
- end index
- path to the file
This information is easy to serialize.

Anyway, you suggest extracting the current code into something like `diplomat-tool-dotnet` (or `diplomat-dotnet`) and make `diplomat-tool` uses it pretty much like it currently does (like a module)?
I agree this is a good initial stopgap. The binary can come later.

**Comment by Manishearth** [comment]
> This is a good thing, are you thinking about embedded `&str` slices into the AST to refer to the original input?

No, probably indexes into the source. Unfortunately proc_macro/proc_macro2's spans aren't stable yet for introspection, adn they don't yet integrate with tools like `miette`. We may need to hack something together.

> Maybe this can be solved if the span holds the following:

yes, that's what "span" means :)

However extracting that info out of proc_macro may not be possible directly and we may need to directly use the `Span` objects, which are not serializable.

> Anyway, you suggest extracting the current code into something like diplomat-tool-dotnet (or diplomat-dotnet) and making diplomat-tool use it pretty much like it currently does (like a module)?

That works really well yeah

**Comment by CBenoit** [comment]
> yes, that's what "span" means :)

Fair enough, I’m sorry about that, my intend was to ensure we were understanding each other because I wasn’t really sure what was the problem here.

> Unfortunately proc_macro/proc_macro2's spans aren't stable yet for introspection, adn they don't yet integrate with tools like `miette`. We may need to hack something together.

> However extracting that info out of proc_macro may not be possible directly and we may need to directly use the `Span` objects, which are not serializable.

I think I understand your concern now.
Indeed, it’s possible to extract some info from `proc_macro2`’s `Span`s by enabling the `span-locations` feature ([`Span::start`](https://docs.rs/proc-macro2/1.0.66/proc_macro2/struct.Span.html#method.start) and [`Span::end`](https://docs.rs/proc-macro2/1.0.66/proc_macro2/struct.Span.html#method.end)), however [`Span::source_file`](https://docs.rs/proc-macro2/latest/proc_macro2/struct.Span.html#method.source_file) is unstable, and since we rely on `syn_inline_mod` to inline all the modules into a single `syn::File` this information is effectively lost.

I’ll back off until this is resolved.

**Comment by Manishearth** [comment]
> Fair enough, I’m sorry about that, my intend was to ensure we were understanding each other because I wasn’t really sure what was the problem here.
> 
> >

Oh, no need to apologize! Not everyone is familiar with terms.

> I’ll back off until this is resolved.

I think it's fine to do the refactoring you sketched in your last comment as a start :smile:. We need dotnet to move to HIR anyway! (also you may find the templating engine useful, see https://github.com/rust-diplomat/diplomat/pull/338) 

Just the general thing of moving to a CLI-based plugin model mediated by JSON doesn't seem too urgent. Having a clear boundary of an HIR being passed across seems like a straightforward way to design this so that that refactoring can be done easily in the future.

**Comment by CBenoit** [comment]
Agreed! :slightly_smiling_face: 

Here is my personal roadmap:
- Move dotnet backend to HIR
- Use `askama` instead of `writeln!` for ad-hoc templating
- Extract the dotnet backend to its own crate that will be depended upon by `diplomat-tool`

I’ll make sure the crate is well-isolated and the API boundary is simple enough:
- HIR + some config as input
- A `FileMap` as output

That way it’ll be easy to wrap into an executable with the appropriate CLI later.

Note for later: exposing a library (as described above) will be useful even when we have a CLI-based plugin. For instance a given backend `diplomat-tool-cpp` could depend on `diplomat-tool-c` as a library because it requires the same `.h` files to be generated, effectively composing both. It’s also nice to let people build their own higher layers with project-specific stuff.

**Comment by Manishearth** [comment]
At this stage I'm not so sure this issue is something I actually want to do. I think HIR should be semi-stable in the long run to support this but I don't think we need a plugin system. It's fine for people to write their own tools wrapping HIR, and we can export common utilities from diplomat-tool.
